Realm/RLMRealm.mm

```diff
@@ -713,6 +713,58 @@ + (BOOL)performMigrationForConfiguration:(RLMRealmConfiguration *)configuration
     return success;
 }
 
+- (void)dispatchAsyncWithBlock:(void(^)(RLMRealm *))block {
+    [self dispatchAsyncWithObjects:@[] block:^(RLMRealm * _Nonnull realm, NSArray<RLMObject *> * _Nonnull) {
+        block(realm);
+    }];
+}
+
+- (void)dispatchAsyncWithObject:(RLMObject *)objectToHandOver block:(void(^)(RLMRealm *, RLMObject *))block {
+    [self dispatchAsyncWithObjects:@[objectToHandOver] block:^(RLMRealm * _Nonnull realm,
+                                                               NSArray<RLMObject *> * _Nonnull singleObjectArray) {
+        block(realm, singleObjectArray[0]);
+    }];
+}
+
+- (void)dispatchAsyncWithObjects:(NSArray<RLMObject *> *)objectsToHandOver
+                           block:(void(^)(RLMRealm *, NSArray<RLMObject *> *))block {
+    [self dispatchAsyncOnQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)
+                   withObjects:objectsToHandOver
+                         block:block];
+}
+
+- (void)dispatchAsyncOnQueue:(dispatch_queue_t)queue
+                 withObjects:(NSArray<RLMObject *> *)objectsToHandOver
+                       block:(void(^)(RLMRealm *, NSArray<RLMObject *> *))block {
+    std::vector<realm::Row> rowsToHandOver;
+    rowsToHandOver.reserve(objectsToHandOver.count);
+    NSMutableArray<NSString *> *classNames = [NSMutableArray arrayWithCapacity:objectsToHandOver.count];
+    for (RLMObject *object in objectsToHandOver) {
+        if (self != object.realm) {
+            @throw RLMException(@"Can only hand over objects from the Realm they belong to.");
+        }
+        rowsToHandOver.push_back(object->_row);
+        [classNames addObject: object.objectSchema.className];
+    }
+    std::shared_ptr<Realm::HandoverPackage> package = _realm->package_for_handover(rowsToHandOver);
+
+    RLMRealmConfiguration *config = self.configuration;
+    dispatch_async(queue, ^{
+        @autoreleasepool {
+            RLMRealm *realm = [RLMRealm realmWithConfiguration:config error:nil];
```

Comment:
How do you suggest we expose this error? I think we need either (1) a second closure for failure or (2) a nullable error argument to the callback in addition to making all other arguments nullable as well.

It seems like (1) is the better option, but it breaks trailing closure syntax in Swift unless we make it the first argument. We could define it like this
```swift
public func async<O: Object>(onQueue queue: DispatchQueue = defaultQueue,
                             handingOver objects: [O],
                             onFailure recover: (ErrorType) -> () = { fatalError(String($0)) },
                             execute block: (Realm, [O]) -> ()) { … }
```
and call it like this
```swift
realm.async(handingOver: [carl, daisy], onFailure: { error in performCustomRecovery(error) }) { realm, objects in
    // do async stuff
}
```

Let me know if you have a more clever solution, @tgoyne or @austinzheng.